### ISSUE TITLE
[aot_eager] fastNLP_Bert - Eager has None grads, while Aot has zero grads

### ISSUE THREAD

**Issue by anijain2305** [pytorch/torchdynamo#1102]
Repro

~~~
import torch
import torch.nn as nn

import torchdynamo
from torchdynamo.utils import same

dont_include = True


def run_fwd_maybe_bwd(gm, args):
    """
    Runs a forward and possibly backward iteration for a given mod and args.
    """
    from torchdynamo.testing import collect_results
    from torchdynamo.testing import reduce_to_scalar_loss
    out = gm(*args)
    loss = reduce_to_scalar_loss(out)
    loss.backward()
    return collect_results(gm, out, loss, args)


class BertPooler(nn.Module):
    def __init__(self):
        super(BertPooler, self).__init__()
        self.dense = nn.Linear(in_features=768, out_features=768, bias=True)

    def graph_break(self, a, b):
        print("graph break")
        return a

    def forward(self, hidden_states):
        a = hidden_states * 2
        b = self.dense(a)
        return self.graph_break(a, b)


x = torch.randn(torch.Size([6, 476, 768]), device="cuda", requires_grad=True)

mod = BertPooler().cuda()
opt_mod = torchdynamo.optimize("aot_nop")(mod)

ref = run_fwd_maybe_bwd(mod, (x,))
res = run_fwd_maybe_bwd(opt_mod, (x,))
assert same(ref, res, tol=0.01)
~~~

For native PyTorch, the backward does not touch `nn.Linear` and therefore gradients of `nn.Linear` are `None`.

For aot_nop, TorchDynamo created graphs has `(a, b)` as output, and then passes it on to Aot Autograd, which does not know that `b` will be later discarded. Therefore, `nn.Linear` grads are all zeros (and not `None`)

cc @Chillee @ezyang

**Comment by anijain2305** [comment]
In this case, the input to the `backward` of `autograd.Function` object is all zeros, when it should have been `None`.

Overall, this seems like a corner case to me. `zero grad` and `None` are basically equivalent.

**Comment by ezyang** [comment]
This is probably got to do with custom autograd function; if the input is not used we still populate zero tensors for it. cc @albanD

**Comment by albanD** [comment]
I think there are two things:
- From the autograd point of view, `None` or a `Tensor full of zeros` is semantically the same. So we can return one or the other.
- If you want to get `None` for missing gradients in custom Function instead of Tensors full of zeros, you can set `ctx.set_materialize_grads(False)`. See doc for more details: https://pytorch.org/docs/stable/notes/extending.html Note that you will need to properly handle the case when grad_output are None.

**Comment by anijain2305** [comment]
Fixed by https://github.com/pytorch/torchdynamo/pull/1124
